Invoke op (#157)
stellar/soroban-docs: docs/learn/stellar-core-and-soroban.mdx

---
sidebar_position: 11

> title: Using Stellar Core to interact with Soroban

Reviewer: The name needs reframing a little. Could you also rename the file to `invoking-contracts-with-transactions`?
```suggestion
title: Invoking Contracts with Stellar Transactions
```

---

> Stellar Core has an new Operation named `InvokeHostFunctionOp` that can be used to interact with
> Soroban. The [Soroban CLI] abstracts these details away from the
> user, but you may find the need to build the XDR transaction and submit it to
> the network directly.
> 
> [Soroban CLI]: ../getting-started/setup#install-the-soroban-cli

Reviewer: This is less about stellar-core, and more about Stellar:
```suggestion
Stellar supports invoking and deploying contracts with a new Operation named `InvokeHostFunctionOp`. The [`soroban-cli`] abstracts these details away from the
user, but SDKs do not yet and if you're building a dapp you'll probably find yourself building the XDR transaction and to submit to
the network.

The `InvokeHostFunctionOp` can be used to invoke host functions. Host functions are built-in to Soroban and give the invoker access to:
- Invoking contract functions.
- Deploying new contracts.
- Deploying token contracts for Stellar Assets.

[`soroban-cli`]: ../getting-started/setup#install-the-soroban-cli
```

## InvokeHostFunctionOp

> The xdr of the operation below can be found

Reviewer: ```suggestion
The XDR of the operation below can be found
```

[here](https://github.com/stellar/stellar-xdr-next/blob/161e2e5b64425a49f9ccfef7f732ae742ed5eec4/Stellar-transaction.x#L486)
```c++
struct InvokeHostFunctionOp
{
    // The host function to invoke

>     HostFunction function;

Reviewer: We need to include the XDR for `HostFunction` here in this section to help people understand what a host function is. While we enumerate the variants in the next section it will be helpful to display the enum here too.

Author: Added the XDR for `HostFunction`.

    // Parameters to the host function
    SCVec parameters;

    // The footprint for this invocation
    LedgerFootprint footprint;
};
```

> ### function

Reviewer: ```suggestion
### Function
```

     salt. 
4. `HOST_FN_CREATE_TOKEN_CONTRACT_WITH_SOURCE_ACCOUNT`
    - This will call the `create_token_from_source_account` host function,
     creating a [built-in token contract] using the stellar source account that submitted the
     transaction. The contract id will be the SHA256 hash of
     [this](https://github.com/stellar/stellar-xdr-next/blob/161e2e5b64425a49f9ccfef7f732ae742ed5eec4/Stellar-transaction.x#L594) `HashIDPreimage`.
   - `parameters` is expected to contain a salt.
5. `HOST_FN_CREATE_TOKEN_CONTRACT_WITH_ASSET`
   - This will call the `create_token_from_asset` host function, creating a
     [built-in token contract] that wraps a classic Stellar asset.
     The contract id will be the SHA256 hash of
     [this](https://github.com/stellar/stellar-xdr-next/blob/161e2e5b64425a49f9ccfef7f732ae742ed5eec4/Stellar-transaction.x#L592) `HashIDPreimage`. `create_token_from_asset` will also initialize the token
     contract, so the user is not expected to call any intialize functions.
   - `parameters` is expected to contain a xdr `Asset`.

[built-in token contract]: ../built-in-contracts/token-contract.mdx

> ### parameters

Reviewer: ```suggestion
### Parameters
```

This `SCVec` contains the parameters that will be passed to the host function that is being invoked.

> ### footprint

Reviewer: ```suggestion
### Footprint
```

The footprint must contain the `LedgerKeys` that will be read and/or written. More
information about the footprint can be found in the advanced section of [interacting with contracts](interacting-with-contracts#storage-footprint-and-preflight).

> 
> :::danger
> InvokeHostFunctionOp is under active-development and is not yet available.
> :::

Reviewer: I think we should drop this danger. Danger is a bit strong, maybe caution would be on par with existing uses. But it is also available on FutureNet, so it may be confusing to say it isn't available at all.
```suggestion
```
